# CAO 
***
## Author: Fionn McCarthy

## Introduction and Background
***

Below shows the libraries imported to carry out the relevant analysis

In [1]:
# Regular expressions 
import re

# Working with HTTP 
import requests as rq

# Data frames 
import pandas as pd

# Date and time library
import datetime as dt 

# For downloading from URl
import urllib.request as urlrq

***
***
# 2021 CAO Data
***
### The link to the CAO points 2021 can be acccessed **[here](http://www.cao.ie/index.php?page=points&p=2021)**.
Over the course of the assignment the data for the 2021 Leaving Certificate points changed format on the website from HTML to Excel format. For teh purpose of this analysis I think it would be good idea to show how both formats can be extracted. In this case I will first show how regular expressions was utilized to extarct the 2021 CAo data in HTML form and then show how the Excel file was converted into daat for the purpose of this analysis.

In [2]:
# Retrieve the CAO points from the URl below
CAO1 = rq.get('http://www2.cao.ie/points/l8.php')

# View the above is okay, 200 means all okay. 
CAO1 

<Response [200]>

### Get the date and time of right now, in order to use for file creation time throughout notebook.
The datetime library is usesd in order to take a back up of the data retrieved from the CAO website.The reason the datetime library is used is to back up the data each time the notebook is run as to not overwrite the data. 

In [3]:
# date and time of right now.
now = dt.datetime.now()

# Formatting now as a string.
nowstring = now.strftime('%Y%m%d_%H%M%S')

### Saving the Original Data Set 

The server advises to utilize charset 'iso-8859-1' in order to decode the but from research this charset is unable to decode \x96 as it is not defined so we use 'cp1252' instead.

In [4]:
# Setting the charset to decode with 
CAO1.encoding = 'cp1252'

In [5]:
# Setting the charset to decode with 
CAO1.encoding = 'cp1252'

# Creating a file path for the data.
path2021_html = 'data/cao2021_' + nowstring + '.html'

# Saving the original html file before going any further. 
with open(path2021_html, 'w') as f:
    f.write(CAO1.text)

### Regular expressions selects the lines in the we want to keep.

In [6]:
# The regular expression is compiled below 
re_expr = re.compile(r'([A-Z]{2}[0-9]{3})(.*)')

In [7]:
# This fucntion will split out the points that had either an '#' or '*' associated with them.
def points_to_array(s):
    portfolio = ''
    if s[0] == '#':
        portfolio = '#'
    random = ''
    if s[-1] == '*':
        random = '*'
    points = ''
    for i in s:
        if i.isdigit():
            points = points + i
    # Return an array with points, portfolio, random        
    return [points, portfolio, random]

In [8]:
# Path creation for the csv file.
path2021_csv = 'data/cao2021_csv_' + nowstring + '.csv'

Looping through lines 

In [9]:
# Keep track of how many courses we process.
no_lines = 0

# Open and write to the cvs file
with open(path2021_csv, 'w') as f:
    # Write a header row for data set
    f.write(','.join(['code', 'title', 'points_r1', 'points_r2']) + '\n')
    # Loop through lines of the response
    for line in CAO1.iter_lines():
        # Decode using cp1252
        dline = line.decode('cp1252')
        # Match only the lines representing courses.
        if re_expr.fullmatch(dline):
            # Add one to the lines counter.
            no_lines = no_lines + 1
            # The course code, first five characters in string
            course_code = dline[:5]
            # The course title
            course_title = dline[7:57].strip()
            # Course points split on one or more spaces after character 60
            course_points = re.split('  +', dline[60:]) 
            if len(course_points) != 2:
                course_points = course_points[:2]
            # join the fields using a comma 
            linesplit = [course_code, course_title, course_points[0], course_points[1]]
            # Rejoin the substrings with commas in between.
            f.write(','.join(linesplit) + '\n')
        
print(no_lines)

949


#### The above count of 949 courses was verified by a manual count that was carried out in visual studio code on original html file.

In [10]:
# Read csv file into 2021 dataframe
df2021 = pd.read_csv(path2021_csv, encoding='cp1252')

### 2021 Data change of Format on CAO website to excel, I will show how I converted this into data to be utlized in the analysis for this assignment.

In [11]:
# Creating a file path for the excel data.
path2021_excel = 'data/cao2021_' + nowstring + '.xlsx'

In [12]:
# https://stackoverflow.com/questions/19602931/basic-http-file-downloading-and-saving-to-disk-in-python
# Save original file
urlrq.urlretrieve("http://www2.cao.ie/points/CAOPointsCharts2021.xlsx", path2021_excel) 

('data/cao2021_20211203_214956.xlsx',
 <http.client.HTTPMessage at 0x261a86121f0>)

In [13]:
# Download and parse the excel spreadsheet
# From examining the excel file I noted that needed to skip the first 10 trows of excel file that is loaded into dataframe
df2021_2 = pd.read_excel('http://www2.cao.ie/points/CAOPointsCharts2021.xlsx', skiprows = 11)

In [14]:
df2021_2

CATEGORY (ISCED Description)  \
0                    Engineering and engineering trades   
1                                                Health   
2                                                Health   
3                       Biological and related sciences   
4                           Business and administration   
...                                                 ...   
1446  Information and Communication Technologies (ICTs)   
1447                                  Personal services   
1448                 Engineering and engineering trades   
1449                                            Welfare   
1450                        Business and administration   

                                  Course Title Course Code R1 Points  \
0              Music and Instrument Technology       AL605       211   
1                          Pharmacy Technician       AL630       308   
2                               Dental Nursing       AL631       311   
3                              Applied Science       AL632       297   
4                                     Business       AL650       AQA   
...                                        ...         ...       ...   
1446                        Creative Computing       WD211       270   
1447           Recreation and Sport Management       WD212       262   
1448  Mechanical and Manufacturing Engineering       WD230       230   
1449        Early Childhood Care and Education       WD231       266   
1450              Business Information Systems       WD232       261   

     R1 Random R2 Points  R2 Random EOS Points EOS Random EOS Midpoints  \
0          NaN        NaN       NaN        211        NaN           319   
1          NaN        NaN       NaN        308        NaN           409   
2          NaN        NaN       NaN        311        NaN           400   
3          NaN        NaN       NaN        297        NaN           454   
4          NaN        AQA       NaN        AQA        NaN           351   
...        ...        ...       ...        ...        ...           ...   
1446       NaN        NaN       NaN        270        NaN           392   
1447       NaN        NaN       NaN        262        NaN           304   
1448       NaN        230       NaN        230        NaN           361   
1449       NaN        NaN       NaN        266        NaN           366   
1450       NaN        261       NaN        261        NaN           329   

      Course Level                                HEI Test/Interview  AVP    v  
0                6    Athlone Institute of Technology            NaN  NaN  NaN  
1                6    Athlone Institute of Technology            NaN  NaN  NaN  
2                6    Athlone Institute of Technology            NaN  NaN  NaN  
3                6    Athlone Institute of Technology            NaN  NaN  NaN  
4                6    Athlone Institute of Technology            NaN  avp  NaN  
...            ...                                ...            ...  ...  ...  
1446             8  Waterford Institute of Technology            NaN  NaN  NaN  
1447             8  Waterford Institute of Technology            NaN  NaN  NaN  
1448             8  Waterford Institute of Technology            NaN  avp  NaN  
1449             8  Waterford Institute of Technology            NaN  NaN  NaN  
1450             8  Waterford Institute of Technology            NaN  avp  NaN  

[1451 rows x 15 columns]

In [15]:
# Check of bottom value of 2021 file is correct
df2021_2.iloc[-1]

CATEGORY (ISCED Description)          Business and administration
Course Title                         Business Information Systems
Course Code                                                 WD232
R1 Points                                                     261
R1 Random                                                     NaN
R2 Points                                                     261
R2 Random                                                     NaN
EOS Points                                                    261
EOS Random                                                    NaN
EOS Midpoints                                                 329
Course Level                                                    8
HEI                             Waterford Institute of Technology
Test/Interview                                                NaN
AVP                                                           avp
v                                                             NaN
Name: 1450

Validated, above is correct when compared to file.

#### Save Pandas Dataframe for 2021 Excel data  

In [16]:
# Creating a file path for the pandas data.
path2021_2_csv = 'data/cao2021_2_' + nowstring + '.csv'

In [17]:
# Save pandas dataframe to folder
df2021_2.to_csv(path2021_2_csv)

***
***
# 2020 CAO Data
***
### The link to the CAO points 2020 can be acccessed **[here](http://www.cao.ie/index.php?page=points&p=2020)**.

#### Save Original 2020 Data

In [18]:
# Creating a file path for the excel data.
path2020_excel = 'data/cao2020_' + nowstring + '.xlsx'

In [19]:
# https://stackoverflow.com/questions/19602931/basic-http-file-downloading-and-saving-to-disk-in-python
# Save original file
urlrq.urlretrieve("http://www2.cao.ie/points/CAOPointsCharts2020.xlsx", path2020_excel) 

('data/cao2020_20211203_214956.xlsx',
 <http.client.HTTPMessage at 0x261a910b970>)

#### Load Spreadhseet using Pandas

In [20]:
# Download and parse the excel spreadsheet
# From examining the excel file I noted that needed to skip the first 10 trows of excel file that is loaded into dataframe
df2020 = pd.read_excel('http://www2.cao.ie/points/CAOPointsCharts2020.xlsx', skiprows = 10)

In [21]:
df2020

CATEGORY (i.e.ISCED description)  \
0                           Business and administration   
1                         Humanities (except languages)   
2                                                  Arts   
3                                                  Arts   
4                                                  Arts   
...                                                 ...   
1459                       Manufacturing and processing   
1460  Information and Communication Technologies (ICTs)   
1461  Information and Communication Technologies (ICTs)   
1462                                  Personal services   
1463                 Engineering and engineering trades   

                                           COURSE TITLE COURSE CODE2  \
0                                International Business        AC120   
1                                          Liberal Arts        AC137   
2      First Year Art & Design (Common Entry,portfolio)        AD101   
3     Graphic Design and Moving Image Design (portfo...        AD102   
4     Textile & Surface Design and Jewellery & Objec...        AD103   
...                                                 ...          ...   
1459                          Manufacturing Engineering        WD208   
1460                       Software Systems Development        WD210   
1461                                 Creative Computing        WD211   
1462                    Recreation and Sport Management        WD212   
1463           Mechanical and Manufacturing Engineering        WD230   

     R1 POINTS R1 Random * R2 POINTS R2 Random*       EOS EOS Random *  \
0          209         NaN       NaN        NaN       209          NaN   
1          252         NaN       NaN        NaN       252          NaN   
2     #+matric         NaN       NaN        NaN  #+matric          NaN   
3     #+matric         NaN       NaN        NaN  #+matric          NaN   
4     #+matric         NaN       NaN        NaN  #+matric          NaN   
...        ...         ...       ...        ...       ...          ...   
1459       188         NaN       NaN        NaN       188          NaN   
1460       279         NaN       NaN        NaN       279          NaN   
1461       271         NaN       NaN        NaN       271          NaN   
1462       270         NaN       NaN        NaN       270          NaN   
1463       253         NaN       NaN        NaN       253          NaN   

     EOS Mid-point  ...  avp    v Column1 Column2 Column3  Column4  Column5  \
0              280  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1              270  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
2         #+matric  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
3         #+matric  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
4         #+matric  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
...            ...  ...  ...  ...     ...     ...     ...      ...      ...   
1459           339  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1460           337  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1461           318  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1462           349  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1463           369  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   

      Column6  Column7  Column8  
0         NaN      NaN      NaN  
1         NaN      NaN      NaN  
2         NaN      NaN      NaN  
3         NaN      NaN      NaN  
4         NaN      NaN      NaN  
...       ...      ...      ...  
1459      NaN      NaN      NaN  
1460      NaN      NaN      NaN  
1461      NaN      NaN      NaN  
1462      NaN      NaN      NaN  
1463      NaN      NaN      NaN  

[1464 rows x 23 columns]

In [22]:
df2020.iloc[-1]

CATEGORY (i.e.ISCED description)          Engineering and engineering trades
COURSE TITLE                        Mechanical and Manufacturing Engineering
COURSE CODE2                                                           WD230
R1 POINTS                                                                253
R1 Random *                                                              NaN
R2 POINTS                                                                NaN
R2 Random*                                                               NaN
EOS                                                                      253
EOS Random *                                                             NaN
EOS Mid-point                                                            369
LEVEL                                                                      8
HEI                                        Waterford Institute of Technology
Test/Interview #                                                         NaN

#### Save Pandas Dataframe

In [23]:
# Creating a file path for the pandas data.
path2020_csv = 'data/cao2020_' + nowstring + '.csv'

In [24]:
# Save pandas datafram to folder
df2020.to_csv(path2020_csv)

***
***
# 2019 CAO Data
***
### The link to the CAO points 2019 can be acccessed **[here](http://www.cao.ie/index.php?page=points&p=2019)**.
The level 8 data and the level 6 and 7 data are held on two different files so I will first manipulate the level 8 data and then the level 6 and 7 data. 
### Level 8 data


#### In order to reproduce the data these are the steps I followed 
1. Download the original level 8 PDF file from the CAO website (link above). 
2. Open the original file using Microsoft Word. 
3. When open in Word, save the document as a word document. 
4. Re-save the Word document for editing.
5. Delete headers and page numbers in footer.
6. Select all in document and copy. 
7. Paste into Notepad++.
8. Remove institution name lines and blank lines.
9. In column headings, replace 'COURSE AND INSTITUTION' with 'Course', save file into folder as a csv file.
10. Change backticks to apostrophes. 
11. Delete tabs at end of lines: 61, 64, 169, 171, 172, 200, 335, 352, 425, 433, 700, 701, 752, 793, 830, 837, 869 and 894. 
12. Remove tab after 'Mid' first line.
13. Remove double tab on line 37.
14. Delete tabs at end of lines: 28, 107, 178, 201, 266, 370, 373, 391, 415, 437, 464, 494, 518, 535, 557, 571, 604, 625, 670, 679, 689, 785, 795
15. This data was then validated by pasting original data into visual studio code and getting the same number of rows.

In [25]:
df2019 = pd.read_csv('data/cao2019_20211130_205230_edited.csv', sep='\t')

In [26]:
df2019

Course Code                                             Course  EOS    Mid
0         AL801    Software Design with Virtual Reality and Gaming  304  328.0
1         AL802               Software Design with Cloud Computing  301  306.0
2         AL803  Software Design with Mobile Apps and Connected...  309  337.0
3         AL805        Network Management and Cloud Infrastructure  329  442.0
4         AL810                                 Quantity Surveying  307  349.0
..          ...                                                ...  ...    ...
925       WD200                                     Arts (options)  221  296.0
926       WD210                       Software Systems Development  271  329.0
927       WD211                                 Creative Computing  275  322.0
928       WD212                    Recreation and Sport Management  274  311.0
929       WD230           Mechanical and Manufacturing Engineering  273  348.0

[930 rows x 4 columns]

In [27]:
# Adding a column to the dataframe specifying the course level, here level 8
df2019['Level'] = 8

In [47]:
df2019

Course Code                                             Course  EOS  \
0         AL801    Software Design with Virtual Reality and Gaming  304   
1         AL802               Software Design with Cloud Computing  301   
2         AL803  Software Design with Mobile Apps and Connected...  309   
3         AL805        Network Management and Cloud Infrastructure  329   
4         AL810                                 Quantity Surveying  307   
..          ...                                                ...  ...   
925       WD200                                     Arts (options)  221   
926       WD210                       Software Systems Development  271   
927       WD211                                 Creative Computing  275   
928       WD212                    Recreation and Sport Management  274   
929       WD230           Mechanical and Manufacturing Engineering  273   

       Mid  Level  
0    328.0      8  
1    306.0      8  
2    337.0      8  
3    442.0      8  
4    349.0      8  
..     ...    ...  
925  296.0      8  
926  329.0      8  
927  322.0      8  
928  311.0      8  
929  348.0      8  

[930 rows x 5 columns]

### Level 6 & 7 data

#### In order to reproduce the data these are the steps I followed 
1. Download the original level 6 & 7 PDF file from the CAO website (link above). 
2. Open the original file using Microsoft Word. 
3. When open in Word, save the document as a word document. 
4. Re-save the Word document for editing.
5. Delete headers and page numbers in footer.
6. Select all in document and copy. 
7. Paste into Notepad++.
8. Remove institution name lines and blank lines.
9. In column headings, replace 'COURSE AND INSTITUTION' with 'Course', save file into folder as a csv file.
10. Remove tab after 'Mid' first line.
11. Delete tabs at end of lines: 31, 67, 69, 104, 108, 124, 157, 158, 174, 220, 224, 225, 228, 249, 294, 295, 328, 331, 365, 366, 410 and 433. 
12. This data was then validated by pasting original data into visual studio code and getting the same number of rows.

In [72]:
df2019_67 = pd.read_csv('data/cao2019_67_20211203_222130_edited.csv', sep='\t')

In [73]:
df2019_67

Course Code                                            Course  EOS    Mid
0         AL600                                   Software Design  205  306.0
1         AL601                              Computer Engineering  196  272.0
2         AL602                            Mechanical Engineering  258  424.0
3         AL604                                 Civil Engineering  252  360.0
4         AL630                               Pharmacy Technician  306  366.0
..          ...                                               ...  ...    ...
456       WD188                               Applied Health Care  206  339.0
457       WD205  Molecular Biology with Biopharmaceutical Science  208  441.0
458       WD206                            Electronic Engineering  191  322.0
459       WD207                            Mechanical Engineering  179  330.0
460       WD208                         Manufacturing Engineering  195  330.0

[461 rows x 4 columns]

#### Load Spreadhseet using Pandas

In [28]:
df2021

code                                              title points_r1  \
0    AL801     Software Design for Virtual Reality and Gaming       300   
1    AL802  Software Design in Artificial Intelligence for...       313   
2    AL803  Software Design for Mobile Apps and Connected ...       350   
3    AL805    Computer Engineering for Network Infrastructure       321   
4    AL810                                 Quantity Surveying       328   
..     ...                                                ...       ...   
944  WD211                                 Creative Computing       270   
945  WD212                    Recreation and Sport Management       262   
946  WD230           Mechanical and Manufacturing Engineering       230   
947  WD231                 Early Childhood Care and Education       266   
948  WD232                       Business Information Systems       261   

    points_r2  
0         NaN  
1         NaN  
2         NaN  
3         NaN  
4         NaN  
..        ...  
944       NaN  
945       NaN  
946       230  
947       NaN  
948       261  

[949 rows x 4 columns]

In [29]:
df2020

CATEGORY (i.e.ISCED description)  \
0                           Business and administration   
1                         Humanities (except languages)   
2                                                  Arts   
3                                                  Arts   
4                                                  Arts   
...                                                 ...   
1459                       Manufacturing and processing   
1460  Information and Communication Technologies (ICTs)   
1461  Information and Communication Technologies (ICTs)   
1462                                  Personal services   
1463                 Engineering and engineering trades   

                                           COURSE TITLE COURSE CODE2  \
0                                International Business        AC120   
1                                          Liberal Arts        AC137   
2      First Year Art & Design (Common Entry,portfolio)        AD101   
3     Graphic Design and Moving Image Design (portfo...        AD102   
4     Textile & Surface Design and Jewellery & Objec...        AD103   
...                                                 ...          ...   
1459                          Manufacturing Engineering        WD208   
1460                       Software Systems Development        WD210   
1461                                 Creative Computing        WD211   
1462                    Recreation and Sport Management        WD212   
1463           Mechanical and Manufacturing Engineering        WD230   

     R1 POINTS R1 Random * R2 POINTS R2 Random*       EOS EOS Random *  \
0          209         NaN       NaN        NaN       209          NaN   
1          252         NaN       NaN        NaN       252          NaN   
2     #+matric         NaN       NaN        NaN  #+matric          NaN   
3     #+matric         NaN       NaN        NaN  #+matric          NaN   
4     #+matric         NaN       NaN        NaN  #+matric          NaN   
...        ...         ...       ...        ...       ...          ...   
1459       188         NaN       NaN        NaN       188          NaN   
1460       279         NaN       NaN        NaN       279          NaN   
1461       271         NaN       NaN        NaN       271          NaN   
1462       270         NaN       NaN        NaN       270          NaN   
1463       253         NaN       NaN        NaN       253          NaN   

     EOS Mid-point  ...  avp    v Column1 Column2 Column3  Column4  Column5  \
0              280  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1              270  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
2         #+matric  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
3         #+matric  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
4         #+matric  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
...            ...  ...  ...  ...     ...     ...     ...      ...      ...   
1459           339  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1460           337  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1461           318  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1462           349  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1463           369  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   

      Column6  Column7  Column8  
0         NaN      NaN      NaN  
1         NaN      NaN      NaN  
2         NaN      NaN      NaN  
3         NaN      NaN      NaN  
4         NaN      NaN      NaN  
...       ...      ...      ...  
1459      NaN      NaN      NaN  
1460      NaN      NaN      NaN  
1461      NaN      NaN      NaN  
1462      NaN      NaN      NaN  
1463      NaN      NaN      NaN  

[1464 rows x 23 columns]

In [30]:
df2019

Course Code                                             Course  EOS  \
0         AL801    Software Design with Virtual Reality and Gaming  304   
1         AL802               Software Design with Cloud Computing  301   
2         AL803  Software Design with Mobile Apps and Connected...  309   
3         AL805        Network Management and Cloud Infrastructure  329   
4         AL810                                 Quantity Surveying  307   
..          ...                                                ...  ...   
925       WD200                                     Arts (options)  221   
926       WD210                       Software Systems Development  271   
927       WD211                                 Creative Computing  275   
928       WD212                    Recreation and Sport Management  274   
929       WD230           Mechanical and Manufacturing Engineering  273   

       Mid  Level  
0    328.0      8  
1    306.0      8  
2    337.0      8  
3    442.0      8  
4    349.0      8  
..     ...    ...  
925  296.0      8  
926  329.0      8  
927  322.0      8  
928  311.0      8  
929  348.0      8  

[930 rows x 5 columns]

***
***
# Joining the three years of data together.
***
## Concat join

In [31]:
courses2021 = df2021[['code', 'title']]
courses2021

code                                              title
0    AL801     Software Design for Virtual Reality and Gaming
1    AL802  Software Design in Artificial Intelligence for...
2    AL803  Software Design for Mobile Apps and Connected ...
3    AL805    Computer Engineering for Network Infrastructure
4    AL810                                 Quantity Surveying
..     ...                                                ...
944  WD211                                 Creative Computing
945  WD212                    Recreation and Sport Management
946  WD230           Mechanical and Manufacturing Engineering
947  WD231                 Early Childhood Care and Education
948  WD232                       Business Information Systems

[949 rows x 2 columns]

In [32]:
courses2020 = df2020[['COURSE CODE2', 'COURSE TITLE']]
courses2020.columns = ['code', 'title']
courses2020

code                                              title
0     AC120                             International Business
1     AC137                                       Liberal Arts
2     AD101   First Year Art & Design (Common Entry,portfolio)
3     AD102  Graphic Design and Moving Image Design (portfo...
4     AD103  Textile & Surface Design and Jewellery & Objec...
...     ...                                                ...
1459  WD208                          Manufacturing Engineering
1460  WD210                       Software Systems Development
1461  WD211                                 Creative Computing
1462  WD212                    Recreation and Sport Management
1463  WD230           Mechanical and Manufacturing Engineering

[1464 rows x 2 columns]

In [33]:
courses2019 = df2019[['Course Code', 'Course']]
courses2019.columns = ['code', 'title']
courses2019

code                                              title
0    AL801    Software Design with Virtual Reality and Gaming
1    AL802               Software Design with Cloud Computing
2    AL803  Software Design with Mobile Apps and Connected...
3    AL805        Network Management and Cloud Infrastructure
4    AL810                                 Quantity Surveying
..     ...                                                ...
925  WD200                                     Arts (options)
926  WD210                       Software Systems Development
927  WD211                                 Creative Computing
928  WD212                    Recreation and Sport Management
929  WD230           Mechanical and Manufacturing Engineering

[930 rows x 2 columns]

In [34]:
# Combine all three years of data 
allcourses = pd.concat([courses2021, courses2020, courses2019], ignore_index=True)
allcourses

code                                              title
0     AL801     Software Design for Virtual Reality and Gaming
1     AL802  Software Design in Artificial Intelligence for...
2     AL803  Software Design for Mobile Apps and Connected ...
3     AL805    Computer Engineering for Network Infrastructure
4     AL810                                 Quantity Surveying
...     ...                                                ...
3338  WD200                                     Arts (options)
3339  WD210                       Software Systems Development
3340  WD211                                 Creative Computing
3341  WD212                    Recreation and Sport Management
3342  WD230           Mechanical and Manufacturing Engineering

[3343 rows x 2 columns]

In [35]:
# Sort values by code 
allcourses.sort_values('code')

code                                     title
175   AC120                    International Business
949   AC120                    International Business
2581  AC120                    International Business
950   AC137                              Liberal Arts
2582  AC137                              Liberal Arts
...     ...                                       ...
2412  WD230  Mechanical and Manufacturing Engineering
946   WD230  Mechanical and Manufacturing Engineering
3342  WD230  Mechanical and Manufacturing Engineering
947   WD231        Early Childhood Care and Education
948   WD232              Business Information Systems

[3343 rows x 2 columns]

In [36]:
# Finds all extra copies of duplicated rows.
allcourses[allcourses.duplicated()]

code                                              title
949   AC120                             International Business
950   AC137                                       Liberal Arts
952   AD102  Graphic Design and Moving Image Design (portfo...
955   AD204                               Fine Art (portfolio)
956   AD211                         Fashion Design (portfolio)
...     ...                                                ...
3338  WD200                                     Arts (options)
3339  WD210                       Software Systems Development
3340  WD211                                 Creative Computing
3341  WD212                    Recreation and Sport Management
3342  WD230           Mechanical and Manufacturing Engineering

[1416 rows x 2 columns]

In [37]:
# Returns a copy of the data frame with duplciates removed.
allcourses.drop_duplicates()

code                                              title
0     AL801     Software Design for Virtual Reality and Gaming
1     AL802  Software Design in Artificial Intelligence for...
2     AL803  Software Design for Mobile Apps and Connected ...
3     AL805    Computer Engineering for Network Infrastructure
4     AL810                                 Quantity Surveying
...     ...                                                ...
3281  TL801         Animation Visual Effects and Motion Design
3282  TL802                            TV, Radio and New Media
3283  TL803                                   Music Technology
3286  TL812                       Computing with Digital Media
3324  WD148                                  Tourism Marketing

[1927 rows x 2 columns]

In [38]:
# Finds all extra copies of duplicated rows.
allcourses[allcourses.duplicated(subset=['code'])]

code                                              title
949   AC120                             International Business
950   AC137                                       Liberal Arts
951   AD101   First Year Art & Design (Common Entry,portfolio)
952   AD102  Graphic Design and Moving Image Design (portfo...
953   AD103  Textile & Surface Design and Jewellery & Objec...
...     ...                                                ...
3338  WD200                                     Arts (options)
3339  WD210                       Software Systems Development
3340  WD211                                 Creative Computing
3341  WD212                    Recreation and Sport Management
3342  WD230           Mechanical and Manufacturing Engineering

[1692 rows x 2 columns]

In [39]:
# Returns a copy of the data frame with duplciates removed - based only on code.
# We must ignore index in order to not bring through index
allcourses.drop_duplicates(subset=['code'], inplace=True, ignore_index=True)

In [40]:
allcourses

code                                              title
0     AL801     Software Design for Virtual Reality and Gaming
1     AL802  Software Design in Artificial Intelligence for...
2     AL803  Software Design for Mobile Apps and Connected ...
3     AL805    Computer Engineering for Network Infrastructure
4     AL810                                 Quantity Surveying
...     ...                                                ...
1646  SG441                              Environmental Science
1647  SG446                                Applied Archaeology
1648  TL803                                   Music Technology
1649  TL812                       Computing with Digital Media
1650  WD148                                  Tourism Marketing

[1651 rows x 2 columns]

### Joining to the Points

In [41]:
# Set the index to the code column.
df2021.set_index('code', inplace=True)
df2021.columns = ['title', 'points_r1_2021', 'points_r2_2021']
df2021

title points_r1_2021  \
code                                                                      
AL801     Software Design for Virtual Reality and Gaming            300   
AL802  Software Design in Artificial Intelligence for...            313   
AL803  Software Design for Mobile Apps and Connected ...            350   
AL805    Computer Engineering for Network Infrastructure            321   
AL810                                 Quantity Surveying            328   
...                                                  ...            ...   
WD211                                 Creative Computing            270   
WD212                    Recreation and Sport Management            262   
WD230           Mechanical and Manufacturing Engineering            230   
WD231                 Early Childhood Care and Education            266   
WD232                       Business Information Systems            261   

      points_r2_2021  
code                  
AL801            NaN  
AL802            NaN  
AL803            NaN  
AL805            NaN  
AL810            NaN  
...              ...  
WD211            NaN  
WD212            NaN  
WD230            230  
WD231            NaN  
WD232            261  

[949 rows x 3 columns]

In [42]:
# Set the index to the code column.
allcourses.set_index('code', inplace=True)

In [43]:
allcourses = allcourses.join(df2021[['points_r1_2021']])
allcourses

title points_r1_2021
code                                                                   
AL801     Software Design for Virtual Reality and Gaming            300
AL802  Software Design in Artificial Intelligence for...            313
AL803  Software Design for Mobile Apps and Connected ...            350
AL805    Computer Engineering for Network Infrastructure            321
AL810                                 Quantity Surveying            328
...                                                  ...            ...
SG441                              Environmental Science            NaN
SG446                                Applied Archaeology            NaN
TL803                                   Music Technology            NaN
TL812                       Computing with Digital Media            NaN
WD148                                  Tourism Marketing            NaN

[1651 rows x 2 columns]

In [44]:
df2020_r1 = df2020[['COURSE CODE2', 'R1 POINTS']]
df2020_r1.columns = ['code', 'points_r1_2020']
df2020_r1

code points_r1_2020
0     AC120            209
1     AC137            252
2     AD101       #+matric
3     AD102       #+matric
4     AD103       #+matric
...     ...            ...
1459  WD208            188
1460  WD210            279
1461  WD211            271
1462  WD212            270
1463  WD230            253

[1464 rows x 2 columns]

In [45]:
# Set the index to the code column.
df2020_r1.set_index('code', inplace=True)
df2020_r1

points_r1_2020
code                
AC120            209
AC137            252
AD101       #+matric
AD102       #+matric
AD103       #+matric
...              ...
WD208            188
WD210            279
WD211            271
WD212            270
WD230            253

[1464 rows x 1 columns]

In [46]:
# Join 2020 points to allcourses.
allcourses = allcourses.join(df2020_r1)
allcourses

title points_r1_2021  \
code                                                                      
AL801     Software Design for Virtual Reality and Gaming            300   
AL802  Software Design in Artificial Intelligence for...            313   
AL803  Software Design for Mobile Apps and Connected ...            350   
AL805    Computer Engineering for Network Infrastructure            321   
AL810                                 Quantity Surveying            328   
...                                                  ...            ...   
SG441                              Environmental Science            NaN   
SG446                                Applied Archaeology            NaN   
TL803                                   Music Technology            NaN   
TL812                       Computing with Digital Media            NaN   
WD148                                  Tourism Marketing            NaN   

      points_r1_2020  
code                  
AL801            303  
AL802            332  
AL803            337  
AL805            333  
AL810            319  
...              ...  
SG441            NaN  
SG446            NaN  
TL803            NaN  
TL812            NaN  
WD148            NaN  

[1651 rows x 3 columns]

---

## References 